### PR TITLE
Feature/atlas lineage v2

### PR DESCRIPTION
**Pull request by kr7ysztof**

### CODE REVIEW

File: src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala [hunk 1]
```diff
@@ -143,33 +189,30 @@ trait LineageHelpers extends EventProducer {
     val userName = userSTS.userName.value
     val clientHost = clientIPAddress.getAddress().get().getHostAddress
     val clientType = lh.clientType.getOrElse("generic")
-    val bucketObject = lh.bucketObject.getOrElse("emptyObject")
+    val bucketObject = lh.bucketObject
     val pseudoDir = lh.pseduoDir.getOrElse(s"${lh.bucket}/")
-    val objectNameFromCopySrc = lh.copySource.getOrElse("")
+    val objectNameFromCopySrc = getObjectName(lh.copySource.getOrElse(""))
     val bucketNameFromCopySrc = lh.copySource.getOrElse("").split("/").head
```
Comment (arempter) [inline src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala]: aws sdk creates  x-amz-copy-source: test/file3.txt from cmd and via sdk it actually creates it with leading slash, that leads to parsing error and empty bucket via sdk

File: src/main/scala/com/ing/wbaa/rokku/proxy/provider/LineageProviderAtlas.scala [hunk 2]
```diff
@@ -58,20 +62,22 @@ trait LineageProviderAtlas extends LineageHelpers {
...
 
         // delete on abort multipart
         // DELETE /ObjectName?uploadId=UploadId
-        case HttpMethods.DELETE if lineageHeaders.queryParams.getOrElse("").contains("uploadId") && !bucketObject.isEmpty =>
+        case HttpMethods.DELETE if lineageHeaders.queryParams.getOrElse("").contains("uploadId") && bucketObject.isDefined =>
           deleteEntityLineage(lineageHeaders.bucketObject.getOrElse(""), userSTS)
 
-        case _ => Future.successful(Done)
+        case _ =>
+          logger.warn("no case for lineage {}", httpRequest)
```
[the remark this answers is not in the document]
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/rokku/proxy/provider/LineageProviderAtlas.scala]: done

File: src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala [hunk 3]
```diff
@@ -96,15 +140,15 @@ trait LineageHelpers extends EventProducer {
     val userName = userSTS.userName.value
     val clientHost = clientIPAddress.getAddress().get().getHostAddress
     val clientType = lh.clientType.getOrElse("generic")
-    val bucketObject = lh.bucketObject.getOrElse("emptyObject")
+    val bucketObject = lh.bucketObject //.getOrElse("emptyObject")
```
[the remark this answers is not in the document]
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala]: done

File: src/main/scala/com/ing/wbaa/rokku/proxy/provider/LineageProviderAtlas.scala [hunk 4]
```diff
@@ -16,37 +17,40 @@ trait LineageProviderAtlas extends LineageHelpers {
...
-          createSingleEntity(lineageHeaders.bucket, userSTS, bucketEntity(lineageHeaders.bucket, userSTS.userName.value, System.nanoTime()))
+        case HttpMethods.PUT if !lineageHeaders.bucket.isEmpty && pseudoDir.isEmpty && bucketObject.isEmpty =>
+          createSingleEntity(lineageHeaders.bucket, userSTS, bucketEntity(lineageHeaders.bucket, userSTS.userName.value, System.nanoTime(), lineageHeaders.classifications.getOrElse(BucketClassification(), List.empty)))
 
         // rm bucket
-        case HttpMethods.DELETE if !lineageHeaders.bucket.isEmpty && bucketObject == "emptyObject" =>
+        case HttpMethods.DELETE if !lineageHeaders.bucket.isEmpty && pseudoDir.isEmpty && bucketObject.isEmpty =>
           deleteEntityLineage(lineageHeaders.bucket, userSTS, AWS_S3_BUCKET_TYPE)
-          deleteEntityLineage(s"${lineageHeaders.bucket}/", userSTS, AWS_S3_PSEUDO_DIR_TYPE) // we also have to remove pseudodir root
+        //deleteEntityLineage(s"${lineageHeaders.bucket}/", userSTS, AWS_S3_PSEUDO_DIR_TYPE) // we also have to remove pseudodir root
```
[the remark this answers is not in the document]
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/rokku/proxy/provider/LineageProviderAtlas.scala]: done

File: src/main/scala/com/ing/wbaa/rokku/proxy/provider/LineageProviderAtlas.scala [hunk 5]
```diff
@@ -16,37 +17,40 @@ trait LineageProviderAtlas extends LineageHelpers {
...
-          val externalObject = s"$EXTERNAL_OBJECT_OUT/${bucketObject.split("/").takeRight(1).mkString}"
+        case HttpMethods.GET if lineageHeaders.queryParams.isEmpty || lineageHeaders.queryParams.contains("encoding-type") && bucketObject.isDefined =>
+          val externalObject = s"$EXTERNAL_OBJECT_OUT/${bucketObject.get.split("/").takeRight(1).mkString}"
           readOrWriteLineage(lineageHeaders, userSTS, Read(), clientIPAddress, Some(externalObject))
 
         // put object from outside of ceph
-        case HttpMethods.PUT if lineageHeaders.queryParams.isEmpty && lineageHeaders.copySource.isEmpty && !bucketObject.isEmpty =>
-          val externalObject = s"$EXTERNAL_OBJECT_IN/${bucketObject.split("/").takeRight(1).mkString}"
+        case HttpMethods.PUT if lineageHeaders.queryParams.isEmpty && lineageHeaders.copySource.isEmpty && (pseudoDir.isDefined || bucketObject.isDefined) =>
+          val externalObject = s"$EXTERNAL_OBJECT_IN/${bucketObject.getOrElse(pseudoDir.get).split("/").takeRight(1).mkString}"
```
[the remark this answers is not in the document]
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/rokku/proxy/provider/LineageProviderAtlas.scala]: done

File: src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala (hunk 1 above)
Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala]: done

File: src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala [hunk 6]
```diff
@@ -46,29 +46,75 @@ trait LineageHelpers extends EventProducer {
...
       extractHeaderOption(httpRequest, "User-Agent").flatMap(extractClient),
       httpRequest.uri.rawQueryString,
       extractHeaderOption(httpRequest, "x-amz-copy-source"),
-      extractHeaderOption(httpRequest, CLASSIFICATIONS_HEADER).map(_.split(",")),
+      extractClassifications(httpRequest),
       extractMetadataHeader(extractHeaderOption(httpRequest, METADATA_HEADER))
     )
   }
 
+  def getBucketName(fullPatch: String): String = fullPatch.split("/").filter(_.nonEmpty).head
```
Comment (arempter) [inline src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala]: typo in fullPatch

Comment (kr7ysztof) [reply src/main/scala/com/ing/wbaa/rokku/proxy/provider/atlas/LineageHelpers.scala]: done

**Pull request merged**